[Ray backend] Better error when pg topology is bad.

rkooo567 opened the pull request:
Closes https://github.com/vllm-project/vllm/issues/6956
Closes https://github.com/vllm-project/vllm/issues/2406

- If the resources of a driver node are not reserved by pg, it raises a RuntimeError
- If the node has less GPUs reserved than tp size, it raise an error (maybe it could be warning log as well. To be discussed). 

Also added some tests
---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

rkooo567 committed:
basic working

rkooo567 committed:
move test to a 2 nodes tests.

rkooo567 committed:
lint fix test

youkaichao reviewed (commented):
can this guarantee the driver worker node is included in the placement group?

youkaichao commented on vllm/executor/ray_utils.py:
```diff
@@ -83,6 +87,49 @@ def assert_ray_available():
+
+    for node_id, bundles in node_id_to_bundle.items():
+        if len(bundles) < parallel_config.tensor_parallel_size:
+            raise RuntimeError(
```
please remove this runtime error. tensor parallel size can be larger than one node GPUs if people use infiniband.

you can make it a warning, saying that it will be slow if you don't have infiniband.

rkooo567 replied:
sgtm

rkooo567 commented:
ah, I misunderstood the requirement.

I think there's a way to get it working. Are we aligned on this semantic? 

- if all tp cannot place on the same node, error out.
- if current node is not included in the pg, error out (and we internally prioritize scheduling on a current node).

rkooo567 commented on vllm/executor/ray_utils.py:
```diff
@@ -144,12 +191,15 @@ def initialize_ray_cluster(
         ray.get(current_placement_group.ready(), timeout=1800)
 
+    assert current_placement_group is not None
+    _verify_bundles(current_placement_group, parallel_config)
```
ah yeah you are right. I misunderstood the comment. Working on prioritizing current node. I think there's no really clean way to do this, but let me figure it out

youkaichao commented:
1. if no pg found, create a pg and **make sure the current node is in the pg**
2. if we already have a pg, but current node is not included in the pg, error out
3. if tp cross nodes, warn

rkooo567 commented:
gotcha. ETA eod today

rkooo567 committed:
Merge branch 'main' into fix-ray-value-error-due-to-missing-pg

rkooo567 committed:
fixed

rkooo567 committed:
fixed

rkooo567 commented:
Addressed comments;

- the first bundle always require node_ip resource, which means the first bundle always has to be created in a current node.
    - there's no API to use node_id here unfortunately in ray now... we are planning to add it, but it will be after oct. 
- Better error message while waiting to create a placement group. 
- Logger.warning if tp workers are not in the same node.

rkooo567 committed:
linting

youkaichao commented on tests/distributed/test_multi_node_topology.py:
```diff
@@ -0,0 +1,75 @@
+
+from vllm.utils import cuda_device_count_stateless
+
+TARGET_TEST_SUITE = os.environ.get("TARGET_TEST_SUITE", "L4")
```
please clean up the tests and remove unnecessary code. this test does not need `TARGET_TEST_SUITE` I think.

youkaichao commented on vllm/executor/ray_utils.py:
```diff
@@ -141,15 +230,32 @@ def initialize_ray_cluster(
+                f"available in a node {current_node_id=} {current_ip=}.")
+        # This way, at least bundle is required to be created in a current
+        # node.
+        placement_group_specs[0][f"node:{current_ip}"] = 0.001
```
is this the key to make sure current node is included in the placement group?

rkooo567 replied:
yes that's right. I failed to find other clean way to support this unfortunately...

rkooo567 commented:
will fix tests and address comments by this sun

rkooo567 committed:
Merge branch 'main' into fix-ray-value-error-due-to-missing-pg

rkooo567 committed:
.

youkaichao commented on tests/distributed/test_multi_node_topology.py:
```diff
@@ -0,0 +1,70 @@
+
+    # Simulate 2 node clusters, 1 GPU each.
+    cluster = Cluster()
+    head_node = cluster.add_node(num_cpus=8, num_gpus=1, resources={"head": 1})
```
why `num_cpus=8` here? is 8 some magic number?

We don't have any guarentee on the number of GPUs we have for this 2 GPUs test.

rkooo567 replied:
ray doesn't require real hardware cpus when specifying resources like this. 8 is just a random number. (I think technically num_cpus=1 should also work)

youkaichao commented on tests/distributed/test_multi_node_topology.py:
```diff
@@ -0,0 +1,70 @@
+@pytest.mark.parametrize("model, distributed_executor_backend, test_suite", [
+    ("facebook/opt-125m", "ray"),
+])
+def test_multi_node_bad_topology(
```
I think you should test it with 2 nodes test (where ray cluster is already created), and try to launch the test with 1 gpus.

rkooo567 replied:
hmm yeah I think we need real 2 nodes if we want to use node ip resources. let me fix it.

youkaichao commented on vllm/executor/ray_utils.py:
```diff
@@ -98,6 +104,88 @@ def assert_ray_available():
+                "tensor_parallel_size=%d "
+                "is smaller than the reserved number of GPUs ({len(bundles)} "
+                "GPUs) in a node %s. Tensor parallel workers can be "
+                "spread out to 2 nodes which can degrade the performance. "
```
"unless you have fast interconnect across nodes, like Infiniband"

rkooo567 committed:
Merge branch 'main' into fix-ray-value-error-due-to-missing-pg

rkooo567 committed:
fixed

rkooo567 committed:
ip

rkooo567 committed:
fixed

rkooo567 committed:
test 2

rkooo567 committed:
fix

youkaichao commented on tests/distributed/test_multi_node_topology.py:
```diff
@@ -0,0 +1,71 @@
+        print(record)
+
+    # Simulate there's no GPU in a current node.
+    @ray.remote(num_gpus=1)
+    class Actor:
+        pass
+
+    # a is created on a head node.
+    actors = [Actor.remote() for _ in range(2)]  # type: ignore
+    ray.get([a.__ray_ready__.remote() for a in actors])
```
is it guaranteed that these actors are allocated in the head node?

rkooo567 replied:
yes. it is a little of implementation detail, but actor is prioritized to be scheduled on a node that creates them.

But good point though. Let me just use https://docs.ray.io/en/master/ray-core/api/doc/ray.util.scheduling_strategies.NodeAffinitySchedulingStrategy.html#ray.util.scheduling_strategies.NodeAffinitySchedulingStrategy so that we don't need to rely on impl details..

rkooo567 committed:
fixed

rkooo567 committed:
fix

rkooo567 commented:
discussed offline. we want to make sure all gpus in the current node is used before other nodes are used. Will modify code and tests accordingly. sorry for the delay

youkaichao committed:
Merge branch 'main' into fix-ray-value-error-due-to-missing-pg

youkaichao committed:
update tests

youkaichao committed:
update tests

youkaichao committed:
refine

youkaichao committed:
fix test

youkaichao committed:
fix test

youkaichao committed:
rearrange tests

darthhexx commented:
These changes have unfortunately broken [Ray 2.9 compatibility](https://github.com/vllm-project/vllm/blob/530821d00cb2beeb8dc62f74f0e4e0003868dc93/requirements-cuda.txt#L5), because `ray._private.state.available_resources_per_node` is named [_available_resources_per_node](https://github.com/ray-project/ray/blob/ray-2.9.3/python/ray/_private/state.py#L747) in 2.9.3 and only named [available_resources_per_node](https://github.com/ray-project/ray/blob/ray-2.10.0/python/ray/_private/state.py#L747) from 2.10 onwards.

Does this mean 2.9 is no longer supported or should we conditionally import the correct class function instead?

rkooo567 commented:
Hmm I guess this never been stable API, and that's probably the issue.

I think we can conditionally import this. @darthhexx would you like to tackle this real quick?

darthhexx commented:
Yip, on it.

darthhexx commented:
Prepped [a PR to conditionally import](https://github.com/vllm-project/vllm/pull/8767), but as mentioned in the PR description we also need the fastapi bump in #8212 to be [pinned to match Ray 2.9.3](https://github.com/vllm-project/vllm/issues/8212#issuecomment-2370349099) to keep Ray 2.9 compatibility.
